### ISSUE TITLE
Reproducible builds

### ISSUE THREAD

**Issue by lstrojny** [symfony/symfony#25958]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | yes
| BC Break report? | -
| RFC?             | -
| Symfony version  | x.y.z

Reproducible builds are all the craze right now (exhibit a: https://reproducible-builds.org/). The basic premise is: code generation, compilation, heck, builds in general should behave deterministically, meaning: given the same set of inputs return the same output.

The main pain point with Symfony comes from the cache warmup process. `console cache:warmup` 
 will return a different results on two consecutive builds. The three main problem areas I've found so far:

 - Container IDs are randomly generated
 - Names and symbols of lazy proxies are randomly generated
 - Compiled twig templates are randomly generated

Before we start digging deeper into this, first question: is that something Symfony wants to properly support or is that something that should somehow live outside of Symfony?

### [Demo project](https://github.com/lstrojny/symfony-reproducible-builds)

**Comment by fabpot** [comment]
I think this is something that needs to be fixed (not saying it will be easy, but this is a sensible goal).

**Comment by nicolas-grekas** [comment]
AFAIK, none of the listed items are randomly named. These names are all based on a hash of their content.

**Comment by stof** [comment]
> Compiled twig templates are randomly generated

That's not true. The name of the compiled class name for Twig templates is a hash of some deterministic input.

**Comment by javiereguiluz** [comment]
The random Twig template names was true in the past in some cases ... but was changed a while ago: https://github.com/twigphp/Twig/pull/2465

**Comment by Tobion** [comment]
proxy class names are not deterministic currently: https://github.com/symfony/symfony/pull/25671#issuecomment-359280579

**Comment by nicolas-grekas** [comment]
@Tobion right I forgot about your comment.
@lstrojny would you like to work on a fix for lazy proxies?

**Comment by lstrojny** [comment]
Here is an example diff between two builds with stock Symfony 3.4: https://gist.github.com/lstrojny/11c61b50ac2a49d3043d55a4ab2d797b

We can see that Twig’s `$__internal` variables are random, the timestamp in the cache files are non-deterministic and the container build time and build id (although I guess that can be remedied already).

**Comment by lstrojny** [comment]
@nicolas-grekas I'll look into the container proxy topic

**Comment by fabpot** [comment]
@lstrojny I think internal variable names can be changed to be deterministic, should be easy to fix.

**Comment by lstrojny** [comment]
The hardest issue for me is `time()` calls, as they are non-deterministic by definition. I wonder if we want to introduce a custom time() function that can be seeded through an environment variable. Something like this:

```php
class Reproducer
{
    private static $reproducibleTime;

    public static function time(): int {
         return self::$reproducibleTime !== null ? self::$reproducibleTime : time(); 
    }

    public static function setReproducibleTime(int $time): void
    {
         self::$reproducibleTime = $time;
    }
}
```

```php
Reproducer::setReproducibleTime(getenv('SYMFONY_REPRODUCIBLE_TIME') ?: null);
```

That way we can dump the cache like this:

```

SYMFONY_REPRODUCIBLE_TIME=0 php app/console cache:warmup
```

**Comment by nicolas-grekas** [comment]
I don't think we want to remove the time - at we don't want to remove the `container.build_time` parameter. If you really want that level of reproducibility, then you should also consider each file's mtime. Which means basically you should run your build in a clock-less container. That's something independent from Symfony.

**Comment by lstrojny** [comment]
@nicolas-grekas I wouldn’t advocate for removing it but for providing a way to set it to something deterministic from the outside like illustrated in the example above. We could replace the the calls to `time()` with a (potentially deterministic) proxy so that in advanced use cases one is able to override what is used for `time()`.

The idea of a clockless container sounds interesting, googling didn’t yield me anything and I have a hard time imagining how Linux would do without any kind of monotonic clock but I am all ears. Do you happen to have a URL for me?

**Comment by javiereguiluz** [comment]
For those interested, on [Symfony Slack chat](https://symfony.com/slack-invite) we discussed a bit about this and here are some comments:

@magarzon:

*"We have several docker containers, and sometimes we see logs from different servers/containers that refer to different compiled twig templates, but that they are actually the same twig template, so I think is a good idea."*

@sroze:

*"That's very good from a security perspective. Everything is kind of going this way, including the low bricks with the immutable infrastructures, read-only containers, etc."*

**Comment by lstrojny** [comment]
To have all resources available here, there is a related composer issue: https://github.com/composer/composer/issues/7049

**Comment by nicolas-grekas** [comment]
@lstrojny AFAIK, only `container.build_time` remains on Symfony's side (Twig is another story.)
Can you work on a PR to make it configurable? It could be just a new `container.force_build_time` parameter that would be read at compile-time to populate `container.build_time`, WDYT?

**Comment by javiereguiluz** [comment]
@nicolas-grekas could you please share some insights about why is the `container.build_time` param needed? I'm just asking because I can't find any usage of it in the Symfony code.

In case we must maintain the `container.build_time` param for some reason ... we could make it deterministic by assigning to it the most recent "last modified time" of all the files processed to build the container.

**Comment by nicolas-grekas** [comment]
`container.build_time` is not used that's true, it's just metadata. But `container.build_id` is, (to construct unique cache prefixes) and is directly derived from the current time.

> we could make it deterministic by assigning to it the most recent "last modified time" of all the files processed

I don't understand: this provides zero difference with how `container.build_time/container.build_id` are built - it's still reliant on current time.

**Comment by sroze** [comment]
I believe that @javiereguiluz's point (with which I agree with) is to replace the dependency of `container.build_*`: instead of using the "current time", it would use the "most recent last modified time of all the files".

Of course, the mtime might not be 100% accurate but it's probably good enough to have a "reproducible build". We just put a hypothesis that in order to consider the "inputs" similar, the files mtimes should be similar.

**Comment by nicolas-grekas** [comment]
> note that file mtimes are not stored in git repos, so mtimes are essentially the time of checkouts

**Comment by sroze** [comment]
> Of course, the mtime might not be 100% accurate but it's probably good enough to have a "reproducible build". We just put a hypothesis that in order to consider the "inputs" similar, the files mtimes should be similar.

😜

**Comment by nicolas-grekas** [comment]
Reproducible builds mean that someone else should be able to compile the full app, take a hash of the compiled artifacts, and get that same as the "original author's". Filemtimes doesn't fit this.

+ the use case we have (constructing unique cache prefixes) doesn't fit, unless we take the filemtime of ALL the files in the app. I don't think we should go into that complex direction at all.

**Comment by javiereguiluz** [comment]
So, if we don't use `container.build_time` except to generate a hash for `container.build_id` ... what if we remove `container.build_time` and generate the `container.build_id` hash with `hash_ini()`, `hash_update()` and `hash_final()` of the files involved in the generation of the container as we do in the GlobResource of the Config component?

**Comment by nicolas-grekas** [comment]
> hash of the files involved in the generation of the container

Because that's not enough. The cache prefix must change if anything changes in the source, all files included, because we don't know how the cache is used, so it could be "sourced" from anywhere in the source.

**Comment by Tobion** [comment]
Monolog handler id are generated using `uniqid` see https://github.com/symfony/monolog-bundle/blob/e60b0130a3fe4ce715c205d8384767c5e681e87a/DependencyInjection/MonologExtension.php#L194
Those are inlined most of the time I guess. But still that conceptually cannot be reproducible either.

**Comment by lstrojny** [comment]
@nicolas-grekas indeed, I want to work on fixating build time from the outside as well as setting the ctime of the cache files via a parameter

**Comment by derrabus** [comment]
@Tobion Looks like the `MonologExtension` could do the inlining already, if I'm not missing something. Should I try a PR?

**Comment by derrabus** [comment]
I've opened symfony/monolog-bundle#248.

**Comment by lstrojny** [comment]
@derrabus cool!

**Comment by lstrojny** [comment]
I've put together [a demo project](https://github.com/lstrojny/symfony-reproducible-builds)  that showcases reproducible builds.

**Comment by nicolas-grekas** [comment]
Is this story over? Shall we close?
A blog post would be great :)

**Comment by lstrojny** [comment]
For now I don’t anything else being on Symfony’s side. The only semi-unresolved issue is https://github.com/ocramius/ProxyManager where it looks like we need to depend on PHP 7.2 (or have people who are more sincerely invested apply the patch).

A blog post would be cool to raise awareness. Anything from my side I can provide for it?

**Comment by javiereguiluz** [comment]
@lstrojny thanks for your great work to make Symfony builds reproducible. The blog post is now published: https://symfony.com/blog/new-in-symfony-reproducible-builds

I'd like to add something about this in the Symfony Docs, so I need to ask you a favor. Could you please show me a simple example of what do you have to configure to make the Symfony build reproducible?

I'm mostly interested in what and how do you configure things like #26127, #26131, #26128 to make your build reproducible. Any other tip/comment/recommendation is greatly appreciated. Thanks!

**Comment by nicolas-grekas** [comment]
@lstrojny about the ProxyManager issue, there is a possibility to fix this on our side: search-and-replace in `Symfony\Bridge\ProxyManager\LazyProxy\PhpDumper::getProxyCode()`, if you want to have a look.

**Comment by lstrojny** [comment]
@nicolas-grekas that’s a nice hack 👍 Let me look into that.

@javiereguiluz have a look at the demo project here: https://github.com/lstrojny/symfony-reproducible-builds

**Comment by lstrojny** [comment]
@nicolas-grekas see #26176

**Comment by nicolas-grekas** [comment]
Thanks for contribution this to Symfony @lstrojny.
